### DISCUSSION TITLE
Enable OpenMp during vcpkg build

### DISCUSSION THREAD

**Discussion by hrkalona** [microsoft/vcpkg#41621] [General]
Hello I am building mpfr by doing `.\vcpkg.exe install mpfr:x64-windows-release`

I am not using the original sources that were defined inside portfile.cmake, but rather one updated variant of mine which is packaged into the original tar.xz. (Some files are updated)

Here is the original portfile: https://github.com/microsoft/vcpkg/blob/master/ports/mpfr/portfile.cmake

```
#vcpkg_download_distfile(ARCHIVE
#    URLS "http://www.mpfr.org/mpfr-${VERSION}/mpfr-${VERSION}.tar.xz" "https://ftp.gnu.org/gnu/mpfr/mpfr-${VERSION}.tar.xz"
#    FILENAME "mpfr-${VERSION}.tar.xz"
#    SHA512 bc68c0d755d5446403644833ecbb07e37360beca45f474297b5d5c40926df1efc3e2067eecffdf253f946288bcca39ca89b0613f545d46a9e767d1d4cf358475
#)

vcpkg_download_distfile(ARCHIVE
    OUT_SOURCE_PATH SOURCE_PATH
    URLS "file:///C:/cygwin64/home/ckalonak/mpfr-4.2.1.tar.xz"
    FILENAME "mpfr-4.2.1.tar.xz"
    SHA512 "2760afad0dbbc098e1d23786dae0d96474355e4dd5d11b9583af0a2d867ab7bede33c697bc68eb29834c24d41e8f7b6a95807fbb17fbea57e2fb2de9a0a8abdd"
)
```
On this updated code I have some openmp directives so I want to somehow force

```
vcpkg_configure_make(
    SOURCE_PATH "${SOURCE_PATH}"
    AUTOCONFIG
)
```
To configure the build in order to use openmp and then build it using:

`vcpkg_install_make()`
I have tried to find a way to do it for hours, and ChatGPT is hallucinating like hell! Anyone has any ideas of how to make the build use the openmp directives?

**Comment by Osyotr** [comment]
If you've added openmp directives to the sources you should've modified the buildsystem too.
For autoconfig it's probably something like this: https://stackoverflow.com/questions/9037434/#comment26922785_9037860

**Reply by hrkalona** [reply]
I will give it a try, but I wanted to have more flexibility and toggle the directives on or off during build.
All the "sources" are compressed to the tar.xz file and then during the vcpkg process they get uncompressed and copied and then the build begins. If I add -fopenmp on the makefile, then this will be part of the tar.xz and I will always have to build with the directives enabled and I loose any chance of having this configurable.

**Reply by dg0yt** [reply]
Well, you can add an option to the build, or you can control the CFLAGS.

**Reply by hrkalona** [reply]
I tried setting the CFLAGS before the configure call, and the following things happen.
1) If I set something to CFLAGS I override whatever configure was going to add.
2) If I set CFLAGS as -fopenmp then because I use msvc it gets ignored because it expects /openmp
3) If I set CFLAGS as /openmp the configure fails (I will post the specific error later on)

In the case that I did something wrong all together, where you suggesting something else when you were mentioning CFLAGS or options? If you have some examples please feel free to tell me so I can try!

**Reply by hrkalona** [reply]
@dg0yt I have rerun my tests:

original configure CFLAGS
configure:19433: WARNING: CFLAGS="-Xcompiler -nologo -Xcompiler -utf-8 -Xcompiler -MP -Xcompiler -MD -Xcompiler -O2 -Xcompiler -Oi -Xcompiler -Gy -Xcompiler -Z7"

1. set(CFLAGS "-fopenmp") before configure
ignored by configure configure:19433: WARNING: CFLAGS="-Xcompiler -nologo -Xcompiler -utf-8 -Xcompiler -MP -Xcompiler -MD -Xcompiler -O2 -Xcompiler -Oi -Xcompiler -Gy -Xcompiler -Z7"

2. set(CFLAGS "-fopenmp") before configure and add option CONFIGURE_ENVIRONMENT_VARIABLES CFLAGS to vcpkg_configure_make
Overriding the other configure cflags configure:19433: WARNING: CFLAGS="-fopenmp", but since I am using microsoft's compiler cl : Command line warning D9002 : ignoring unknown option '-fopenmp'
Running a benchmark on my code has the same execution time as a sigle threaded run, so I guess -fopenmp was ignored

set(CFLAGS "/openmp") before configure and add option CONFIGURE_ENVIRONMENT_VARIABLES CFLAGS to vcpkg_configure_make
The configure fails on its compilation test steps
cl : Command line warning D9024 : unrecognized source file type 'C:/Users/ckalonak/Documents/vcpkg/downloads/tools/msys2/212eb2bc95cd8a82/openmp', object file assumed

Any ideas?

**Reply by dg0yt** [reply]
> 1. set(CFLAGS "-fopenmp") before configure

Add `"CFLAGS=\$CFLAGS -fopenmp"` to the OPTIONS of `vcpkg_configure_make`.

> 2. add option CONFIGURE_ENVIRONMENT_VARIABLES CFLAGS 

Well, CONFIGURE_ENVIRONMENT_VARIABLES might work with MSVC...

**Reply by hrkalona** [reply]
I changed  vcpkg_configure_make to this:
`vcpkg_configure_make(
    SOURCE_PATH "${SOURCE_PATH}"
    AUTOCONFIG
    OPTIONS "CFLAGS=\$CFLAGS -fopenmp" 
)`

But I am getting cl : Command line warning D9002 : ignoring unknown option '-fopenmp'

I also tested this:

`vcpkg_configure_make(
    SOURCE_PATH "${SOURCE_PATH}"
    AUTOCONFIG
    OPTIONS "CFLAGS=\$CFLAGS /openmp" 
)`

But I am getting this:
cl : Command line warning D9024 : unrecognized source file type 'C:/Users/ckalonak/Documents/vcpkg/downloads/tools/msys2/212eb2bc95cd8a82/openmp', object file assumed

Its the same thing I mentioned before.

**Reply by hrkalona** [reply]
I have also tried to disregard the configuration on the vcpkg side and add

CFLAGS += /openmp
LDFLAGS += /openmp

Inside the Makefile.am (inside src)
The code builds but the my pragmas are disregarded:


          #pragma omp parallel sections num_threads(2)

           {

               #pragma omp section
               {
                   mpfr_add(temp1, re, im, rnd_mode); // temp1 = re + im
                   mpfr_sub(temp2, re, im, rnd_mode); // temp2 = re - im
                   mpfr_mul(temp2, temp1, temp2, rnd_mode); // temp2 = (re + im) * (re - im)
               }

               #pragma omp section
               {
                   mpfr_mul(temp3, re, im, rnd_mode); // temp3 = re * im
                   mpfr_mul_2ui(temp3, temp3, 1, rnd_mode); // temp3 = 2 * re * im
               }

           }



This code is still executed on a single thread fashion. 
By the way I have built the same code without vcpkg on Linux using -fopenmp and my pragmas work, the performance also reflects that.

**Reply by dg0yt** [reply]
IIUC `-fopenmp` isn't understood by MSVC.

`/openmp` is documented but ... `vcpkg_configure_make` on Windows needs MSYS `sh`, and the leading `/` triggers a heuristical POSIX-to-Windows filepath conversion when calling `cl` (Windows!) from `sh` (MSYS!).
Solution: MSVC should also understand `-openmp`. Or use MSYS2's environment variables to control conversion.

If you need to target different platforms and compilers, you must write cmake code to select and pass the corresponding option.

**Reply by hrkalona** [reply]
Thank you, OPTIONS "CFLAGS=\$CFLAGS -openmp"  works!

**Comment by hrkalona** [comment]
Should I create an issue out of this discussion? I really need to find a way to make this work.

**Reply by dg0yt** [reply]
No.
